src/wp-includes/interactivity-api/interactivity-api.php

```diff
@@ -0,0 +1,136 @@
+<?php
+/**
+ * Interactivity API: Functions and hooks
+ *
+ * @package WordPress
+ * @subpackage Interactivity API
+ * @since 6.5.0
+ */
+
+/**
+ * Processes the directives on the rendered HTML of the interactive blocks.
+ *
+ * This processes only one root interactive block at a time because the
+ * rendered HTML of that block contains the rendered HTML of all its inner
+ * blocks, including any interactive block. It does so by ignoring all the
+ * interactive inner blocks until the root interactive block is processed.
+ *
+ * @since 6.5.0
+ *
+ * @param array $parsed_block The parsed block.
+ * @return array The same parsed block.
+ */
+function wp_interactivity_process_directives_of_interactive_blocks( array $parsed_block ): array {
+	static $root_interactive_block = null;
+
+	/*
+	 * Checks whether a root interactive block is already annotated for
+	 * processing, and if it is, it ignores the subsequent ones.
+	 */
+	if ( null === $root_interactive_block ) {
+		$block_name = $parsed_block['blockName'];
+		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
+
+		if ( isset( $block_name ) && isset( $block_type->supports['interactivity'] ) && $block_type->supports['interactivity'] ) {
+			// Annotates the root interactive block for processing.
+			$root_interactive_block = array( $block_name, md5( serialize( $parsed_block ) ) );
+
+			/*
+			 * Adds a filter to process the root interactive block once it has
+			 * finished rendering.
+			 */
+			$process_interactive_blocks = static function ( $content, $parsed_block ) use ( &$root_interactive_block, &$process_interactive_blocks ) {
+				// Checks whether the current block is the root interactive block.
+				list($root_block_name, $root_block_md5) = $root_interactive_block;
+				if ( $root_block_name === $parsed_block['blockName'] && md5( serialize( $parsed_block ) ) === $root_block_md5 ) {
```

Comment:
Question: Does the `parsed_block` really need to be serialized and hashed? In PHP, arrays are not passed by reference, so you can do deep comparisons with `===` of two separate identical arrays and get `true`:

```php
<?php

$a = array(
    'name' => 'foo',
    'props' => array( 1, 2, array( 'a', 'b', 'c' ) )
);

$b = array(
    'name' => 'foo',
    'props' => array( 1, 2, array( 'a', 'b', 'c' ) )
);


$c = array(
    'name' => 'foo',
    'props' => array( 1, 2, array( 'A', 'B', 'C' ) )
);

var_dump( $a === $b ); // => bool(true)
var_dump( $a === $c ); // => bool(false)
```

So I think this can be made more efficient by eliminating the serializing and hashing:

```suggestion
			$root_interactive_block = array( $block_name, $parsed_block ) );

			/*
			 * Adds a filter to process the root interactive block once it has
			 * finished rendering.
			 */
			$process_interactive_blocks = static function ( $content, $parsed_block ) use ( &$root_interactive_block, &$process_interactive_blocks ) {
				// Checks whether the current block is the root interactive block.
				list( $root_block_name, $root_parsed_block ) = $root_interactive_block;
				if ( $root_block_name === $parsed_block['blockName'] && $parsed_block === $root_parsed_block ) {
```

Reply:
Fantastic. Thank you Weston, we've been struggling to figure out the best way to do that identification.

Fixed in https://github.com/WordPress/wordpress-develop/pull/5953/commits/553e8a4db9ae12626cdf9a3a6a472d16c7de881d.

cc: @c4rl0sbr4v0